zapr.go

```diff
@@ -154,18 +157,372 @@ func (zl *zapLogger) handleFields(lvl int, args []interface{}, additional ...zap
 		if !isString {
 			// if the key isn't a string, DPanic and stop logging
 			if zl.panicMessages {
-				zl.l.WithOptions(zap.AddCallerSkip(1)).DPanic("non-string key argument passed to logging, ignoring all later arguments", zap.Any("invalid key", key))
+				zl.l.WithOptions(zap.AddCallerSkip(1)).DPanic("non-string key argument passed to logging, ignoring all later arguments", zapAny("invalid key", key))
 			}
 			break
 		}
 
-		fields = append(fields, zap.Any(keyStr, val))
+		fields = append(fields, zapAny(keyStr, val))
 		i += 2
 	}
 
 	return append(fields, additional...)
 }
 
+// TODO: move to logr
+//
+// UseReflection is an optional interface that values in a key/value list
+// may implement. That this interface is implemented provides a hint to
+// loggers that all fields in the value are meant to be logged via
+// reflection even when the logger normally would do something else,
+// like using the Stringer interface.
+//
+// Loggers are not required to support it. Depending on their purpose
+// or configuration, they might still prefer to use the Stringer interface,
+// for example when the goal is to provide human-readable output.
+type UseReflection interface {
+	// LogWithReflection will never get called and can be empty.
+	// It's existence already indicates the intent to prefer reflection
+	// for logging over alternatives like String.
+	LogWithReflection()
```

Comment:
I was considering whether this should be `LogWithReflection() bool` and then have the logger actually call this to determine whether reflection is preferred. At first glance that seems more natural and could be used to override the choice when inlining some type which has a `LogWithReflection() {return true}` with `LogWithReflection() {return false}`.

But that has a performance impact also when the interface is not implemented at all. `zapAny` below can only be written like that when the method doesn't have to be called. The alternatives (wrapping `zap.Any` as in `zapAny2` and individual if checks as in `zapAny3`) are slower.

So, benchmark time....

`zapAny` with the additional interface assertion is slower than `zap.Any` for types that don't have the interface. It is faster when the interface exists because that short-circuits the switch:
```
name              old time/op  new time/op  delta
ZapAny_kmeta-36   37.3ns ± 1%  47.8ns ± 1%  +27.93%  (p=0.008 n=5+5)
ZapAny_kmeta2-36  38.5ns ± 0%  18.9ns ± 0%  -50.84%  (p=0.008 n=5+5)
ZapAny_int-36     29.4ns ± 0%  35.3ns ± 0%  +20.01%  (p=0.008 n=5+5)
ZapAny_bool-36    19.7ns ± 0%  25.4ns ± 3%  +28.93%  (p=0.008 n=5+5)
```

Bool and int have different overhead in all of these comparisons because bool gets checked first.

Extending the switch via copy-and-paste of `zap.Any` is the fastest solution. Wrapping `zap.Any` (= `zapAny2`) is slower for most types:
```
name              old time/op  new time/op  delta
ZapAny_kmeta-36   47.8ns ± 1%  53.8ns ± 0%  +12.54%  (p=0.008 n=5+5)
ZapAny_kmeta2-36  18.9ns ± 0%  18.3ns ± 0%   -3.50%  (p=0.008 n=5+5)
ZapAny_int-36     35.3ns ± 0%  42.9ns ± 1%  +21.57%  (p=0.008 n=5+5)
ZapAny_bool-36    25.4ns ± 3%  32.0ns ± 0%  +25.65%  (p=0.008 n=5+5)
```

If we wanted to call `LogWithReflection` and check the result, we'd have to such wrapping or use consecutive if switches (= `zapAny3`). That's really ugly code and worse for some types (but not all?!) than `zapAny`:
```
name              old time/op  new time/op  delta
ZapAny_kmeta-36   47.8ns ± 1%  73.2ns ± 3%  +53.25%  (p=0.008 n=5+5)
ZapAny_kmeta2-36  18.9ns ± 0%  18.5ns ± 0%   -2.31%  (p=0.008 n=5+5)
ZapAny_int-36     35.3ns ± 0%  38.9ns ± 0%  +10.14%  (p=0.008 n=5+5)
ZapAny_bool-36    25.4ns ± 3%  24.2ns ± 0%   -5.00%  (p=0.008 n=5+5)
```

Reply:
But obviously this is just benchmarking a tiny piece of the overall logging.  To put this into perspective, we are talking about nanoseconds here. A full `Info` call which writes int, bool and KMeta to `io.Discard` is taking 3.7 microseconds and the slowdown caused by replacing `zap.Any` is barely measurable:
```
name     old time/op  new time/op  delta
Info-36  3.70µs ± 3%  3.77µs ± 4%   ~     (p=0.421 n=5+5)
```

Reply:
So overall, perhaps `LogWithReflection() bool` and `zapAny2` are the preferred choice because it's the cleaner API, avoids forking `zap.Any` and the performance difference is neglectable?